src/dev/reg_bank.hh

```diff
@@ -956,45 +975,34 @@ class RegisterBank : public RegisterBankBase
         if (it == _offsetMap.end() || it->first > addr)
             it--;
 
-        if (it->first < addr) {
-            RegisterBase &reg = it->second.get();
-            // Skip at least the beginning of the first register.
-
-            // Figure out what parts of it we're accessing.
-            const off_t reg_off = addr - it->first;
-            const size_t reg_bytes = std::min(reg.size() - reg_off,
-                                              bytes - done);
-
-            // Actually do the access.
-            reg.read(ptr, reg_off, reg_bytes);
-            done += reg_bytes;
-            it++;
-
-            // Was that everything?
-            if (done == bytes)
-                return;
+        std::ostringstream ss;
+        while (done != bytes) {
+          RegisterBase &reg = it->second.get();
+          const size_t reg_off = addr - it->first;
+          const size_t reg_size = reg.size() - reg_off;
+          const size_t reg_bytes = std::min(reg_size, bytes - done);
+
+          if (reg_bytes != reg.size()) {
+              if (_debug_flag) {
+                  ccprintf(ss, "Read register %s, byte offset %d, size %d\n",
+                          reg.name(), reg_off, reg_bytes);
+              }
+              reg.read(ptr + done, reg_off, reg_bytes);
+          } else {
+              if (_debug_flag) {
+                  ccprintf(ss, "Read register %s\n", reg.name());
+              }
+              reg.read(ptr + done);
+          }
+
+          done += reg_bytes;
+          addr += reg_bytes;
+          ++it;
         }
 
-        while (true) {
-            RegisterBase &reg = it->second.get();
-
-            const size_t reg_size = reg.size();
-            const size_t remaining = bytes - done;
-
-            if (remaining == reg_size) {
-                // A complete register read, and then we're done.
-                reg.read(ptr + done);
-                return;
-            } else if (remaining > reg_size) {
-                // A complete register read, with more to go.
-                reg.read(ptr + done);
-                done += reg_size;
-                it++;
-            } else {
-                // Skip the end of the register, and then we're done.
-                reg.read(ptr + done, 0, remaining);
-                return;
-            }
+        if (_debug_flag) {
+            ::gem5::trace::getDebugLogger()->dprintf_flag(
+                curTick(), name(), _debug_flag->name(), "%s", ss.str());
         }
```

Comment:
I would consider using DPRINTFV which has been introduced exactly for this reason..

https://github.com/gem5/gem5/blob/stable/src/base/trace.hh#L189

Reply:
Hi! DPRINTFV expects a flag instance however the _debug_flag here is a pointer that might be nullptr. So here we still need an extra null check for this use case. Do you think we should add yet another macro to deal with this use case to include it within GEM5_UNLIKELY?

Reply:
Yeah you are right. I don't want to stall this any further... Either we use references instead of pointers or we add the pointer version as you suggested. This can be done with another PR (but only if you feel like)

Reply:
I decided not to change this in this PR.